frontends/smv_node.h

```diff
@@ -10,30 +11,1276 @@
 #include "exceptions.h"
 #include "smt-switch/smt.h"
 
+using namespace std;
 /* node to help construct syntax tree in the parser */
 namespace pono {
+class module_node;
 struct SMVnode
 {
+ public:
   enum Nodetype
   {
     TermNode,
     SortNode
   };
+  enum Vartype
+  {
+    BasicT,
+    ModuleT
+  };
+  enum NodeMtype
+  {
+    DEFINE,
+    CONSTANTS,
+    ASSIGN,
+    IVAR,
+    VAR,
+    FROZENVAR,
+    INIT,
+    TRANS,
+    INVAR,
+    INVARSPEC
+  };
+  enum Type
+  {
+    Signed,
+    Unsigned,
+    Integer,
+    Real,
+    Boolean,
+    WordArray,
+    Default
+  };
   smt::Term tm;
   smt::Sort st;
   Nodetype nt;
+  Type bvt;
+  NodeMtype mt;
+  Vartype vt;
+  Type ele_type;
+  SMVnode()
+  {
+    nt = SortNode;
+    bvt = Default;
+    ele_type = Default;
+  }
   SMVnode(smt::Sort s)
   {
     nt = SortNode;
     st = s;
+    bvt = Default;
+    ele_type = Default;
   }
   SMVnode(smt::Term t)
   {
     nt = TermNode;
     tm = t;
+    bvt = Default;
+    ele_type = Default;
+  }
+  SMVnode(smt::Sort s, Type type)
+  {
+    nt = SortNode;
+    st = s;
+    bvt = type;
+    ele_type = Default;
+  }
+  SMVnode(smt::Term t, Type type)
+  {
+    nt = TermNode;
+    tm = t;
+    bvt = type;
+    ele_type = Default;
+  }
+  SMVnode(smt::Term t, Type type, Type element)
+  {
+    nt = TermNode;
+    tm = t;
+    bvt = type;
+    ele_type = element;
   }
+
   Nodetype getNodeType() { return nt; }
+  NodeMtype getNodeMType() { return mt; }
+  Vartype getVarType() { return vt; }
+  Type getType() { return bvt; }
+  Type getElementType() { return ele_type; }
   smt::Sort getSort() { return st; }
   smt::Term getTerm() { return tm; }
+  virtual std::string getName() { return ""; }
+  virtual void generate_ostream(
+      std::string name,
+      std::string prefix,
+      std::unordered_map<string, module_node *> module_list,
+      std::unordered_map<string, string> new_prefix,
+      ostream & s){};
 };  // struct SMVNode
-}  // namespace pono
\ No newline at end of file
+
+class element_node : public SMVnode
+{
+  std::vector<SMVnode *> pa_li;
+
+ public:
+  element_node() {}
+  virtual void preprocess(
+      std::string module,
+      std::string prefix,
+      std::unordered_map<std::string, module_node *> module_list,
+      std::unordered_map<string, string> * new_prefix,
+      ostream & s){};
+  virtual std::vector<SMVnode *> get_list() { return pa_li; }
+};
+
+class module_node
+{
+  std::string module_name;
+  std::string par_name;
+  std::vector<string> par_li;
+  unordered_map<string, SMVnode *> new_par;
+  element_node * var_li;
```

Comment:
Just to check my understanding: these `element_node` pointers each store a vector of `SMVNode *` that contain the corresponding type. As in, `var_li` has a vector of `SMVNode *`'s that correspond to variables. Is that right? Could you add a comment about that if so?